Added option to exclude c14n tranformation from `Transform` XML node

vo-va opened the pull request:
Some software can uses hard restrictions on Transform XML node, that prohibit include c14n transformation into Transform XML node.

Pull request to resolve issue #43

vo-va committed:
Added option to exclude c14n tranformation from `Transform` XML node

Some software can uses hard restrictions on Transform XML node, that prohibit include c14n transformation into Transform XML node.

vo-va committed:
Lines were too long, shortened to 120 characters

vo-va committed:
Removed trailing whitespace

vo-va committed:
Changed travis config to upgrade requirements for OSx

vo-va committed:
Changed travis config again, to upgrade requirements for OSx

vo-va committed:
Continue debug travis file to pass tests on osx

vo-va committed:
Changed travis config file to pass tests on osx

brew return error if package already installed - changed scripts to check package installation.
For python3 use venv module instead virtualenv

kislyuk commented:
Thank you for your contribution. Generally, I am unwilling to add options to the main codebase that result in signatures that don't pass validation by other XML Signature implementations (such as https://github.com/lsh123/xmlsec and https://github.com/yaronn/xml-crypto). In this case I don't have enough information to gather whether or not that is the case, so I have to ask you to give me some details about which XML signature implementations are able to verify the resulting signature. Similarly, I'd like to understand whether setting this option to `False` results in a signature that complies with the XML Signature standard and passes validation with the XML Signature schema (not Microsoft's custom schema). Can you provide information on both of these questions?

Aside from that, there are a couple of practical issues with the PR:
- Tests are required that exercise the option that you introduced.
- The docstring contains a few grammatical errors, it would be great to correct them.

This library is built to be extensible. Even if this option is not introduced into the library, you have the option of subclassing the `XMLSigner` class and overloading the `_build_sig` method.

vo-va committed:
Changed tests to check include_c14n_transform option

vo-va commented:
There is a project https://github.com/vo-va/signxml-test/tree/master with Docker container that verify signed document with xmlsec and xml-crypto. I used xmlsec1 util and example from xml-crypto docs to verify signature. 

> Similarly, I'd like to understand whether setting this option to False results in a signature that complies with the XML Signature standard and passes validation with the XML Signature schema (not Microsoft's custom schema). 

If I understand correctly xmlsign tests verify signed document with this scheme https://github.com/XML-Security/signxml/blob/master/signxml/schemas/xmldsig-core-schema.xsd 
I changed tests code to run test for cases with option in both states, True or False. Therefore I think signature without c14n transformation complies with standard xsd. 

Also standard scheme is not limit amount of transformation https://github.com/XML-Security/signxml/blob/master/signxml/schemas/xmldsig-core-schema.xsd#L117

I am not sure what my changes will test all cases. Can you look at my changes and say is it enough or not?

vo-va committed:
Fixing grammatical errors

kislyuk commented:
Thank you for your patience. I want you to know that I haven't forgotten about this PR, but currently have no time to review it (and the nature of the changes is such that this requires a substantial review). I'll get back to this ASAP, hopefully sometime around next weekend.

bobdoah commented:
Any chance of getting this merged? I too am hitting the same issue.

vo-va committed:
Sync with upstream/master (git@github.com:XML-Security/signxml.git)

vo-va committed:
Updated osx_image, added commands to install virtualenv for python2

Default osx_image in travis uses OS X 10.12.6 (and Xcode 8.3.3) has pip 9.0.1, it fails on install/download packages on OSX, see notes for pip release 9.0.2 https://github.com/pypa/pip/releases/tag/9.0.2
osx_image=xcode9.4 include pip 10.0.1

tjeb commented:
I was about to make a pull request of almost the same modification I had on my local instance, due to other implementations of refusing to accept documents with the c14n transform element included (since it results in multiple Transform elements).

The relevant piece of the specification appears to be here: https://www.w3.org/TR/xmldsig-core2/#sec-Transforms

> Each Reference MUST contain the Transforms element, and this MUST contain one and only one Transform element with an Algorithm of "http://www.w3.org/2010/xmldsig2#transform". This signals the 2.0 syntax and processing (Compatibility mode transforms are described in section B.5 "Compatibility Mode" Transforms and Processing Model).

The canonicalization is specified by an earlier element: https://www.w3.org/TR/xmldsig-core2/#sec-CanonicalizationMethod

> CanonicalizationMethod is a required element that specifies the canonicalization algorithm applied to the SignedInfo element prior to performing signature calculations. This element uses the general structure for algorithms described in section 3.2.1 XML Signature 2.0 Algorithm Identifiers and Implementation Requirements. Implementations MUST support the REQUIRED canonicalization algorithms.

So, add my name to the list of people wanting to have this (or any solution for this problem) merged :)

kislyuk commented:
Thank you for your work on this PR so far, @vo-va, and sorry about the delays in handling this.

I don't currently use software that is incompatible with the way SignXML produces signatures. It's clear that there is demand for fixing the underlying issue here, which as far as I know is that SignXML only produces "Compatibility Mode" signatures, and not "2.0" signatures (thank you @tjeb for pointing toward that fact). I have described the issue in further detail in #142.

With apologies for not having had time to look at this issue in this level of detail sooner, I must ask that someone familiar with these applications write a principled solution for #142, with a separate test case, if you want this issue to be resolved.

I'm going to leave this PR open as a reference, but it has major issues that prevent it from being merged. The PR:

- does not document the underlying incompatibility in sufficient detail
- does not reference software or test cases that demonstrate the incompatibility
- is too vague in the documentation that it does provide
- solves the problem in too narrow a way
- disrupts the existing test cases unnecessarily (the test for this should ideally be a new, separate test case)

A solution to this issue should address these problems.

Again, apologies for the delays in handling this, but I am one person maintaining this library in my spare time outside my line of work, and it carries serious security requirements. Handling issues in this library requires substantial time, focus and clarity in understanding the underlying problems, given the issues in the underlying standard and the diversity of software interfacing with it. If you want to help, please contribute clear, maintainer-friendly solutions.

davidcolclazier commented:
Might I inquire as to the current progress for this PR? Are you still looking for a more robust PR, or have you already implemented this? I notice there are now conflicts for this merge.
